### PR TITLE
Recordable property required by pynn

### PR DESCRIPTION
**Pull request by Christian-B**

### CODE REVIEW

**Comment by Christian-B** [comment]
population.celltype.recordable is something the pynn documentation says a users can call so that is why we want the property.
See http://neuralensemble.org/docs/PyNN/neurons.html
Section: Recording variables and retrieving recorded data
===

I played with several different options.

1. Use a chain of calls to super to build up the list.

2. Keep a list in the Recordable of the variables,
2a. Fill this list during init calls
2b. Fill this list during new calls

3. Have an Abstract static propery that is implemented several times over

4. Use hasattributes to look for getSpikes, getV ect  . IE Ducktyping

===

Rowleys idea should also work. I will have a play with that,

Clearly the method and property names have to be as defined by pynn otherwise this whole thing is not very useful.

**Comment by Christian-B** [comment]
Did two different versions using mro().

Which Alan does not like as he does not know mro()

Clearly when we pick one implementation the others have to be removed.

File: spynnaker/pyNN/models/common/recordable.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+class Recordable(object):
+    """ Indicates that spikes can be recorded from this object
+    """
+
+    __slots__ = ()
+
+    # Here we have to choose between using @abstractstatic or @staticmethod
+    # Using both either fails or cause @abstractstatic to be ignored
+    # However even abstractstatic only enforces that one of the
+    # abstract_x_recordable interfaces implements the method not all.
```
Comment (rowleya) [inline spynnaker/pyNN/models/common/recordable.py]: This is definitely an issue - I don't think python provides any way around this (Java doesn't know how to do this either ;) )

File: spynnaker/pyNN/models/common/recordable.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        if not isinstance(aclass, type):
+            aclass = type(aclass)
+        variables = list()
+        for sub in aclass.mro():
+            if sub in Recordable.__subclasses__():
+                variables.append(sub.get_recordable_variable())
+        return variables
+
+    @staticmethod
+    def get_all_possible_recordable_variables1(aclass):
```
Comment (rowleya) [inline spynnaker/pyNN/models/common/recordable.py]: I think this one is the most correct, as it uses issubclass.

**Comment by Christian-B** [comment]
An alternative approach may be to push the list of recordable variables higher up the stack.
Then instead of using the existance of the various abstract_XYZ_recordable interfaces to do all data recording (except Spikes) with a more general approach.

**Comment by rowleya** [comment]
I think this is how it was before these changes...

**Comment by Christian-B** [comment]
Using a single class for all but spike data type is the better solution.

This stuff is dead but branch left temporarily for reference.

**Pull request closed without merging**